text/000x-nrd-kernels.md

```diff
@@ -0,0 +1,324 @@
+
+* Introduce NRD kernel to transaction, compensate with additional kernel.
+* Introduce NRD kernel to transaction, compensate with kernel offset.
+
+----
+
+#### Payment Channel Implementation
+
+NRD kernels can be used to delay alternate "branches" of conflicting transactions, enabling a payment channel implementation.
+
+A payment channel is represented as a single multi-party output. Each channel state transition is represented as a pair of "close" and "settle" transactions with an NRD kernel enforcing a delay between them. Funds are held in an intermediate multi-party output while delayed. The NRD kernel is reused across both transactions by adjusting kernel offsets.
+
+_X -> Y, K<sub>nrd_a</sub>_ \
+_Y -> [Z<sub>a</sub>, Z<sub>b</sub>], K<sub>nrd_a</sub>_
+
+Alice closes the channel _X_ with their "close" transaction. After a delay Alice can "settle" the funds out to Alice and Bob.
+
+_Attribution_ of "close" and "settle" transactions for each channel state is provided through endpoint specific NRD kernels. This allows the other party to "revoke" old invalid state without the NRD delay.
+
+Each channel state transition involves a new pair of "close" and "settle" transactions for each participant along with a shared "revoke" transaction. The "revoke" transaction simply spends funds back to the channel output and a plain kernel suffices.
+
+_[Z<sub>a</sub>, Z<sub>b</sub>] -> X, K<sub>rev</sub>_
+
+Alice attempts to close old invalid state (_Y<sub>1</sub>_):
+
+_X -> Y<sub>1</sub>, K<sub>nrd_a1</sub>_
+
+Bob can immediately revoke and close current state (_Y<sub>1</sub> -> Y<sub>2</sub>_):
+
+_Y<sub>1</sub> -> ~[Z<sub>a</sub>, Z<sub>b</sub>]~, K<sub>nrd_b1</sub>_ \
+_~[Z<sub>a</sub>, Z<sub>b</sub>]~ -> ~X~, K<sub>rev_1</sub>_ \
+_~X~ -> Y<sub>2</sub>, K<sub>nrd_b2</sub>_ \
+_=> Y<sub>1</sub> -> Y<sub>2</sub>, [K<sub>nrd_b1</sub>, K<sub>rev_1</sub>, K<sub>nrd_b2</sub>]_
+
+Bob publishes only the final cut-through multi-kernel transaction (_Y<sub>1</sub> -> Y<sub>2</sub>_). Bob's individual settle transaction is not revealed.
+
+Neither party can self-revoke without introducing the NRD delay. The other party always has the opportunity to revoke first. Self-revocation cannot be used to lock funds up indefinitely.
+
+----
+
+#### Rollout/Deployment (HF3)
+
+The following rules will be enforced during rollout as part of HF3 -
+
+__Assumptions:__
+
+1. HF3 will occur at height 786,240.
+2. Blocks at height >= 786,240 will have block version >= 4.
+
+__Block Specific Rules:__
+
+1. Blocks containing NRD kernel(s) are only be valid if block version >= 4.
+2. Blocks will be subject to NRD relative lock height rules.
+
+__Transaction Specific Rules:__
+
+1. Transactions containing NRD kernel(s) will not be relayed or broadcast unless the current chain head version >= 4.
+2. Transaction selection from the txpool when building a candidate block, must not select any transactions containing NRD kernel(s) unless the block version >= 4.
+
+# Drawbacks
```

Comment:
As mentioned on keybase, it's probably worth adding that NRD kernels cannot be verified in isolation which makes it a bit harder to verify in parallel due to their kernel reference window. Not impossible though